### PR TITLE
Feature: Display Dishes

### PR DESCRIPTION
**Pull request by aniansson**
## Pivotal: 
https://www.pivotaltracker.com/story/show/137256347

## Changes proposed in this pull request:
Created dish model in database and generated controller for dish. Plus route for page. Creates a global variable for all dishes and passes them to the dish view where all of them are displayed. Plus made Rspec test.

## What I have learned working on this feature: 
Repetition on creating and generating models, making Rspec test. So refreshing for mind. 
Also if using the humanize method, use small letters when creating new items in database.
# Ready for review

### CODE REVIEW

**Review by tochman — COMMENTED** [review]
Good job so far. Will review when time comes. 😜

File: features/display_dishes.feature [hunk 1]
```diff
@@ -0,0 +1,17 @@
+Feature: As a visitor,
+  In order to see example of dishes available in general,
+  I need to be able to see a selection of dishes
+
+Background:
+  Given the following dishes exists
+    |    name   |                        description                              |  price  | readytime | portions |
+    | meatballs |    homecooked with love, including mashed potatoes and sauce    |    49   |   18:00   |    10    |
+    |   taco    |    really spicy authentic Mexican tacos                         |    59   |   16:00   |    10    |
```
Comment (AmberWilkie) [inline features/display_dishes.feature]: No one in this country is making authentic tacos, but not a blocker.

File: config/routes.rb [hunk 2]
```diff
@@ -1,3 +1,5 @@
 Rails.application.routes.draw do
+  get 'dishes/show'
```
Comment (AmberWilkie) [inline config/routes.rb]: Should this really be the show page? We normally use that for individual dishes. This is the kind of info we would ordinarily put in an index for dishes. So url.com/dishes takes you to a list of all dishes and when you want to view an individual dish, it's url.com/dishes/1 - and that's what the `show` method is for.

Comment (AmberWilkie) [reply config/routes.rb]: Either way, there's better Rails syntax for this. we can do:
`resources :dishes, only: [:show]` and that should give you the same routes you have now, plus a helper.

**Review by AmberWilkie — CHANGES_REQUESTED** [review]
Think about this show vs. index thing. If Thomas told you to use show and you're planning to change it later or not provide any individual dish information, no problem.

File: app/models/dish.rb [hunk 3]
```diff
@@ -0,0 +1,7 @@
+class Dish < ApplicationRecord
+  validates :name, presence: true
+  validates :description, presence: true
+  validates :price, presence: true
+  validates :readytime, presence: true
+  validates :portions, presence: true
```
Comment (tochman) [inline app/models/dish.rb]: I would suggest using
```
validates_presence_of : name, :description, etc...
```
= less code. What do you think?

File: app/views/dishes/index.html.erb [hunk 4]
```diff
@@ -0,0 +1,11 @@
+<ul>
+    <% @dishes.each do |dish| %>
+      <li>
+      <%= dish.name.humanize %><br/>
+      <%= dish.description.humanize %>
+      Price: <%= dish.price %>kr
+      Ready for pick-up at <%= dish.readytime.strftime("%H:%M") %>
+      <%= dish.portions %> portions left
+    </li>
```
Comment (tochman) [inline app/views/dishes/index.html.erb:9]: Mind the code indentation...

File: features/display_dishes.feature (hunk 1 above)
Comment (tochman) [reply features/display_dishes.feature]: Also, I can't see that you are testing for the presence of `tacos`. If not used, why is it added to the db`

File: features/step_definitions/basic_steps.rb [hunk 5]
```diff
@@ -0,0 +1,13 @@
+Given(/^the following dishes exists$/) do |table|
+  table.hashes.each do |hash|
+    FactoryGirl.create(:dish, hash)
+  end
+end
+
+When(/^I am on the landing page$/) do
```
Comment (tochman) [inline features/step_definitions/basic_steps.rb]: Is the `dishes_path` really the `landing page`? It seems to conflict with the story the other team is working on.

File: features/step_definitions/basic_steps.rb [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+  visit dishes_path
+end
+
+Then(/^I should see "([^"]*)"$/) do |content|
+  expect(page).to have_content content
+end
```
Comment (tochman) [inline features/step_definitions/basic_steps.rb]: I always prefer to separate the step definition that are assertions into a separate file for better readability.

File: spec/factories/dishes.rb [hunk 7]
```diff
@@ -0,0 +1,9 @@
+FactoryGirl.define do
+  factory :dish do
+    name "MyString"
+    description "MyText"
+    price 1
+    readytime "2017-01-10 16:32:43"
```
Comment (tochman) [inline spec/factories/dishes.rb]: Not a blocker but I would be mindful on how this value is set. I would prefer `Date.today (or .now or whatever..)` Also, I would consider validating that the `readytime` is in the future. (I also have a problem with the attribute itself, `readytime` does not feel like a good name. I would at least separate the words with an underscore `ready_time`)

**Review by AmberWilkie — APPROVED** [review]
Looks great!

File: spec/models/dish_spec.rb [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+
+  describe "DB table" do
+    it { is_expected.to have_db_column :name }
+    it { is_expected.to have_db_column :description }
+    it { is_expected.to have_db_column :price }
+    it { is_expected.to have_db_column :readytime }
+    it { is_expected.to have_db_column :portions }
+  end
+
+  describe "Validation" do
```
Comment (tochman) [inline spec/models/dish_spec.rb]: `Validation` or `Validations`? Are we testing one or several validations?

File: spec/models/dish_spec.rb [hunk 9]
```diff
@@ -0,0 +1,26 @@
+require 'rails_helper'
+
+RSpec.describe Dish, type: :model do
+
+  describe "DB table" do
```
Comment (tochman) [inline spec/models/dish_spec.rb]: Are we testing one or several tables?

**Review by tochman — CHANGES_REQUESTED** [review]

File: features/step_definitions/basic_steps.rb (hunk 5 above)
Comment (Philippoes) [reply features/step_definitions/basic_steps.rb]: We are changing this when landing page is merged. We are adding the dishes as a partial to the landing page. That's why it says landing

File: features/display_dishes.feature (hunk 1 above)
Comment (Philippoes) [reply features/display_dishes.feature]: Yes, we totally missed that. Fixing asap.

File: features/step_definitions/basic_steps.rb (hunk 5 above)
Comment (tochman) [reply features/step_definitions/basic_steps.rb]: Okay, thank you for the clarification.

File: app/views/dishes/index.html.erb [hunk 10]
```diff
@@ -0,0 +1,11 @@
+<ul>
+  <% @dishes.each do |dish| %>
+    <li>
+      <%= dish.name.humanize %><br/>
+      <%= dish.description.humanize %>
+      Price: <%= dish.price %>kr
+      Ready for pick-up at <%= dish.ready_time.strftime("%H:%M") %>
+      <%= dish.portions %> portions left
+    </li>
```
Comment (tochman) [inline app/views/dishes/index.html.erb:9]: Still not formatted correctly. There's a auto format tool in RubyMine.

File: spec/models/dish_spec.rb [hunk 11]
```diff
@@ -0,0 +1,26 @@
+require 'rails_helper'
+
+RSpec.describe Dish, type: :model do
+
+  describe "DB tables" do
+    it { is_expected.to have_db_column :name }
+    it { is_expected.to have_db_column :description }
+    it { is_expected.to have_db_column :price }
+    it { is_expected.to have_db_column :ready_time }
```
Comment (tochman) [inline spec/models/dish_spec.rb:9]: You need to add a migration to rename that attribute in order to make these tests to pass.

**Review by tochman — APPROVED** [review]

File: features/step_definitions/display_dishes_steps.rb [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+
+When(/^I am on the "([^"]*)" page$/) do |page|
+  if page == "landing"
+    visit root_path
+  end
+
+  if page == "Meatballs"
+    id = Dish.find_by(name: "meatballs").id
+    visit "/dishes/#{id}"
+  end
```
Comment (tochman) [inline features/step_definitions/display_dishes_steps.rb:15]: Please use a case statement here. Much cleaner

**Review by tochman — CHANGES_REQUESTED** [review]

**Review by tochman — APPROVED** [review]

**Pull request merged**